network/ip_tracker.go

```diff
@@ -0,0 +1,360 @@
+	// false positives, we limit each validator to maxIPEntriesPerValidator in
+	// the bloom filter.
+	bloomAdditions map[ids.NodeID]int // Number of IPs added to the bloom
+	bloomSalt      []byte
+	maxBloomCount  int
+}
+
+func (i *ipTracker) ManuallyTrack(nodeID ids.NodeID) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	if !i.validators.Contains(nodeID) {
+		i.onValidatorAdded(nodeID)
+	}
+	i.manuallyTracked.Add(nodeID)
+}
+
+func (i *ipTracker) WantsConnection(nodeID ids.NodeID) bool {
+	i.lock.RLock()
+	defer i.lock.RUnlock()
+
+	return i.validators.Contains(nodeID)
+}
+
+func (i *ipTracker) ShouldVerifyIP(ip *ips.ClaimedIPPort) bool {
+	i.lock.RLock()
+	defer i.lock.RUnlock()
+
+	if !i.validators.Contains(ip.NodeID) {
+		return false
+	}
+
+	prevIP, ok := i.mostRecentValidatorIPs[ip.NodeID]
+	return !ok || // This would be the first IP
+		prevIP.Timestamp < ip.Timestamp // This would be a newer IP
+}
+
+// AddIP returns true if the addition of the provided IP updated the most
+// recently known IP of a validator.
+func (i *ipTracker) AddIP(ip *ips.ClaimedIPPort) bool {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	if !i.validators.Contains(ip.NodeID) {
+		return false
+	}
+
+	prevIP, ok := i.mostRecentValidatorIPs[ip.NodeID]
+	if !ok {
+		// This is the first IP we've heard from the validator, so it is the
+		// most recent.
+		i.updateMostRecentValidatorIP(ip)
+		// Because we didn't previously have an IP, we know we aren't currently
+		// connected to them.
+		return true
+	}
+
+	if prevIP.Timestamp >= ip.Timestamp {
+		// This IP is not newer than the previously known IP.
+		return false
+	}
+
+	i.updateMostRecentValidatorIP(ip)
+	i.removeGossipableIP(ip.NodeID)
+	return true
+}
+
+func (i *ipTracker) GetIP(nodeID ids.NodeID) (*ips.ClaimedIPPort, bool) {
+	i.lock.RLock()
+	defer i.lock.RUnlock()
+
+	ip, ok := i.mostRecentValidatorIPs[nodeID]
+	return ip, ok
+}
+
+func (i *ipTracker) Connected(ip *ips.ClaimedIPPort) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	i.connected[ip.NodeID] = ip
+	if !i.validators.Contains(ip.NodeID) {
+		return
+	}
+
+	prevIP, ok := i.mostRecentValidatorIPs[ip.NodeID]
+	if !ok {
+		// This is the first IP we've heard from the validator, so it is the
+		// most recent.
+		i.updateMostRecentValidatorIP(ip)
+		i.addGossipableIP(ip)
+		return
+	}
+
+	if prevIP.Timestamp > ip.Timestamp {
+		// There is a more up-to-date IP than the one that was used to connect.
+		return
+	}
+
+	if prevIP.Timestamp < ip.Timestamp {
+		i.updateMostRecentValidatorIP(ip)
+	}
+	i.addGossipableIP(ip)
+}
+
+func (i *ipTracker) Disconnected(nodeID ids.NodeID) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	delete(i.connected, nodeID)
+	i.removeGossipableIP(nodeID)
+}
+
+func (i *ipTracker) OnValidatorAdded(nodeID ids.NodeID, _ *bls.PublicKey, _ ids.ID, _ uint64) {
+	i.lock.Lock()
+	defer i.lock.Unlock()
+
+	i.onValidatorAdded(nodeID)
+}
+
+func (i *ipTracker) onValidatorAdded(nodeID ids.NodeID) {
+	if i.manuallyTracked.Contains(nodeID) {
```

Comment:
If we are manually tracking a NodeID and it is a validator, I'm confused why we wouldn't still add its IP as gossipable?